Store capability OpenAPI V3 JSON schema in ConfigMap (#1106)
oam-dev/kubevela: docs/en/platform-engineers/openapi-v3-json-schema.md

# OpenAPI v3 JSON Schema of a definition

## Why is OpenAPI v3 JSON Schema of a definition needed?

In definitions, [parameter](https://kubevela.io/#/en/platform-engineers/workload-type?id=_4-define-template) section are expected to be set by developers when creating `Application` object.
While there is another GUI way for developers to input all parameter fields by rendering `parameter` section to [OpenAPI v3 Specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#format).

> For example, we can generate OpenAPI v3 JSON Schema from `parameter` of Workload Type [webservice](https://kubevela.io/#/en/developers/references/workload-types/webservice).

Reviewer: ```suggestion
For example,  after you created a definition object, KubeVela will generate OpenAPI v3 JSON Schema from `parameter` of Workload Type [webservice](https://kubevela.io/#/en/developers/references/workload-types/webservice).
```

						"properties": {
							"secretKeyRef": {
								"description": "Selects a key of a secret in the pod's namespace",
								"properties": {
									"key": {
										"description": "The key of the secret to select from. Must be a valid secret key",
										"title": "key",
										"type": "string"
									},
									"name": {
										"description": "The name of the secret in the pod's namespace to select from",
										"title": "name",
										"type": "string"
									}
								},
								"required": ["name", "key"],
								"title": "secretKeyRef",
								"type": "object"
							}
						},
						"required": ["secretKeyRef"],
						"title": "valueFrom",
						"type": "object"
					}
				},
				"required": ["name"],
				"type": "object"
			},
			"title": "env",
			"type": "array"
		},
		"image": {
			"description": "Which image would you like to use for your service",
			"title": "image",
			"type": "string"
		},
		"port": {
			"default": 80,
			"description": "Which port do you want customer traffic sent to",
			"title": "port",
			"type": "integer"
		}
	},
	"required": ["image", "port"],
	"type": "object"
}
```

> Rendering the schema by [form-render](https://github.com/alibaba/form-render) or [React JSON Schema form](https://github.com/rjsf-team/react-jsonschema-form), a web form will show as below.

Reviewer: ```suggestion
You can render the schema by [form-render](https://github.com/alibaba/form-render) or [React JSON Schema form](https://github.com/rjsf-team/react-jsonschema-form). A web form can be as below.
```

Author: 好像没什么区别。

Reviewer: 需要一个主语，不然不通顺

![](../../resources/json-schema-render-example.jpg)

> ## How to get the JSON Schema of a definition

Reviewer: ```suggestion
## How does the JSON Schema of a definition generated
```

Author: 我目的是想告诉用户怎么获取。

Reviewer: Then：

```
## How to use the generated JSON Schema of definition
```

When a platform builder applies a WorkloadDefinition or TraitDefinition, a ConfigMap will be created in a namespace same

> as the definition (currently it's `vela-system` as definition are not namespaced) , and the name is in the format

Reviewer: ```suggestion
as the definition (by default it will be in `vela-system` namespace) , and the name is in the format
```

`schema-$definitionName`. The key of ConfigMap data is `openapi-v3-json-schema`.

For example, we can use the following command to get the JSON Schema of `webservice`.

```shell

> $ kubectl get cm schema-webservice -n vela-system -o yaml

Reviewer: please add a label on the configmap, so we can list all configmaps generated by KubeVela.

```
$ kubectl get cm -n vela-system -l definition.oam.dev=schema
NAME               DATA   AGE
schema-webservice   1       2d
schema-worker         1       1d
```

Author: Good point!

    a key of a secret in the pod''s namespace","properties":{"key":{"description":"The
    key of the secret to select from. Must be a valid secret key","title":"key","type":"string"},"name":{"description":"The
    name of the secret in the pod''s namespace to select from","title":"name","type":"string"}},"required":["name","key"],"title":"secretKeyRef","type":"object"}},"required":["secretKeyRef"],"title":"valueFrom","type":"object"}},"required":["name"],"type":"object"},"title":"env","type":"array"},"image":{"description":"Which
    image would you like to use for your service","title":"image","type":"string"},"port":{"default":80,"description":"Which
    port do you want customer traffic sent to","title":"port","type":"integer"}},"required":["image","port"],"type":"object"}'
kind: ConfigMap
metadata:
  name: schema-webservice
  namespace: vela-system
  ownerReferences:
  - apiVersion: core.oam.dev/v1alpha2
    blockOwnerDeletion: true
    controller: true
    kind: WorkloadDefinition
    name: webservice
    uid: e1321d3a-6537-4f81-be23-77082f3cdede
  resourceVersion: "80216641"
  selfLink: /api/v1/namespaces/vela-system/configmaps/schema-webservice

>   uid: 2b9cc399-7f93-42d8-a8e2-e2743775fd6f

Reviewer: you should remove these details in doc

Author: 就是完全不展示 configmap，还是简化展示？我已经把一些无关紧要的字段删掉了。
